sdk/messaging/azservicebus/receiver.go

```diff
@@ -333,118 +348,114 @@ func (r *Receiver) receiveMessagesImpl(ctx context.Context, maxMessages int, opt
-func (r *Receiver) drainLink(ctxForLoggingOnly context.Context, receiver internal.AMQPReceiver, messages []*ReceivedMessage) ([]*ReceivedMessage, error) {
-	// start the drain asynchronously. Note that we ignore the user's context at this point
-	// since draining makes sure we don't get messages when nobody is receiving.
-	if err := receiver.DrainCredit(context.Background()); err != nil {
-		tab.For(ctxForLoggingOnly).Debug(fmt.Sprintf("Draining of credit failed. link will be closed and will re-open on next receive: %s", err.Error()))
-
-		// if the drain fails we just close the link so it'll re-open at the next receive.
-		if err := r.amqpLinks.Close(context.Background(), false); err != nil {
-			tab.For(ctxForLoggingOnly).Debug(fmt.Sprintf("Failed to close links on ReceiveMessages cleanup. Not fatal: %s", err.Error()))
-		}
-	}
+		for {
+			amqpMessage, err := lwid.Receiver.Receive(ctx)
 
-	// Draining data from the receiver's prefetched queue. This won't wait for new messages to
-	// arrive, so it'll only receive messages that arrived prior to the drain.
-	for {
-		am, err := receiver.Prefetched(context.Background())
+			if err != nil {
+				return err
+			}
 
-		if am == nil || internal.IsCancelError(err) {
-			break
-		}
+			all = append(all, newReceivedMessage(amqpMessage))
+			got++
 
-		if err != nil {
-			// something fatal happened, we will just
-			_ = r.amqpLinks.Close(context.TODO(), false)
+			if got == credits {
+				// no excess credits on link
+				needsDrain = false
+				break
+			}
 
-			if len(messages) > 0 {
-				return messages, nil
-			} else {
-				return nil, err
+			if cancel == nil {
+				// replace the context that we're using for everything with a new one that will cancel
+				// after a period of time.
+				ctx, cancel = context.WithTimeout(ctx, defaultTimeAfterFirstMessage)
+				defer cancel()
 			}
 		}
 
-		messages = append(messages, newReceivedMessage(ctxForLoggingOnly, am))
+		return nil
+	}, utils.RetryOptions(r.retryOptions))
+
+	ret := func(err error) ([]*ReceivedMessage, error) {
+		if len(all) > 0 {
+			// we don't return the error here because we did retrieve _some_ messages and you can still
+			// use them.
+			return all, nil
+		} else {
+			return nil, err
+		}
 	}
 
-	return messages, nil
-}
-
-// getMessages receives messages until a link failure, timeout or the user
-// cancels their context.
-func (r *Receiver) getMessages(ctx context.Context, receiver internal.AMQPReceiver, maxMessages int, maxWaitTimeAfterFirstMessage time.Duration) ([]*ReceivedMessage, error) {
-	var messages []*ReceivedMessage
-
-	for {
-		var amqpMessage *amqp.Message
-		amqpMessage, err := receiver.Receive(ctx)
+	if err != nil && !internal.IsCancelError(err) {
+		return ret(err)
+	}
 
-		if err != nil {
-			if internal.IsCancelError(err) {
-				return messages, nil
+	if needsDrain {
+		// start the drain asynchronously. Note that we ignore the user's context at this point
```

Comment:
Are we sure it's safe to ignore the user's context?  Can this take too long?

Reply:
All we're really waiting for here is an ACK from the service so it should be fast.